spec/unit/helper_rake_methods_spec.rb

```diff
@@ -0,0 +1,355 @@
+        subject.run_build
+      end
+
+      it "does not check for go_availbility" do
+        expect(subject).not_to receive(:go_available?)
+        subject.run_build
+      end
+    end
+
+    context "if go environment unavailable" do
+      before(:each) do
+        expect(subject).to receive(:go_available?).and_return(false)
+      end
+
+      it "returns false" do
+        expect(subject.run_build).to be(false)
+      end
+
+      it "does not build" do
+        expect(subject).not_to receive(:run_go_build)
+        subject.run_build
+      end
+    end
+
+    context "if run in git and revision has changed" do
+      before(:each) do
+        allow(subject).to receive(:runs_in_git?).and_return(true)
+      end
+
+      it "returns false if build fails" do
+        expect(subject).to receive(:run_go_build).and_return(false)
+        expect(subject.run_build).to be(false)
+      end
+
+      it "writes go version file" do
+        version_file = File.join(helper_dir, "version.go")
+        expect(File.exist?(version_file)).to be(false)
+        expect(subject.run_build).to be(true)
+        expect(File.exist?(version_file)).to be(true)
+      end
+
+      it "creates git revision file after successful build" do
+        git_revision_file = File.join(helper_dir, "..", ".git_revision")
+        expect(File.exist?(git_revision_file)).to be(false)
+        expect(subject).to receive(:run_go_build).and_return(true)
+        expect(subject.run_build).to be(true)
+        expect(File.exist?(git_revision_file)).to be(true)
+      end
+
+      it "does not create git revision file after failed build" do
+        git_revision_file = File.join(helper_dir, "..", ".git_revision")
+        expect(subject).to receive(:run_go_build).and_return(false)
+        expect(subject.run_build).to be(false)
+        expect(File.exist?(git_revision_file)).to be(false)
+      end
+    end
+
+    context "if run in git and revision hasn't changed" do
+      before(:each) do
+        allow(subject).to receive(:runs_in_git?).and_return(true)
+        # git revision file is compared to current GIT head
+        subject.write_git_revision_file
+      end
+
+      it "returns false if build fails" do
+        expect(subject).to receive(:run_go_build).and_return(false)
+        expect(subject.run_build).to be(false)
+      end
+
+      it "builds go binary if source files are newer than binary" do
+        FileUtils.touch(File.join(helper_dir, "machinery-helper"), :mtime => Time.now - 10)
+        FileUtils.touch(File.join(helper_dir, "machinery_helper.go"))
+        expect(subject).to receive(:run_go_build).and_return(true)
+        expect(subject.run_build).to be(true)
+      end
+
+      it "does not build go binary if source files are older than binary" do
+        FileUtils.touch(File.join(helper_dir, "machinery-helper"))
+        FileUtils.touch(File.join(helper_dir, "machinery_helper.go"), :mtime => Time.now - 10)
```

Comment:
Use the new Ruby 1.9 hash syntax.<br>Do not use `Time.now` without zone. Use one of `Time.zone.now`, `Time.now.current`, `Time.now.in_time_zone`, `Time.now.utc`, `Time.now.getlocal`, `Time.now.iso8601`, `Time.now.jisx0301`, `Time.now.rfc3339`, `Time.now.to_i`, `Time.now.to_f` instead.